File: crates/hm-dsl-engine/harmont-py/harmont/_secret.py
"""Secret references for the Harmont pipeline SDK.

`hm.secrets["NAME"]` returns a SecretRef — a *reference* to a secret, never its
value. The value is resolved at run time (locally from .env + the process
environment; in the cloud from the org/pipeline secret store). Use it anywhere
an env value is accepted: `hm.sh("deploy", env={"TOKEN": hm.secrets["DEPLOY_TOKEN"]})`.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_SECRET_NAME = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


@dataclass(frozen=True)
class SecretRef:
    """A reference to a stored secret, identified by name."""

    name: str

    def __repr__(self) -> str:  # never leaks a value — there is none to leak
        return f"SecretRef({self.name!r})"


class _Secrets:
    """Subscript accessor exposed as `hm.secrets`."""

    def __getitem__(self, name: str) -> SecretRef:
        if not isinstance(name, str) or not _SECRET_NAME.match(name):
            msg = (
                f"secret name {name!r} is invalid: names must match "
                r"[A-Za-z_][A-Za-z0-9_]* (letters, digits, underscores; no leading digit)."
            )
            raise ValueError(msg)
        return SecretRef(name)


secrets = _Secrets()

File: crates/hm-dsl-engine/harmont-ts/src/secret.ts
const SECRET_NAME = /^[A-Za-z_][A-Za-z0-9_]*$/;
const SECRET_BRAND = Symbol("harmont.secretRef");

/** A reference to a stored secret, identified by name. Never holds a value. */
export interface SecretRef {
  readonly [SECRET_BRAND]: true;
  readonly name: string;
}

export function isSecretRef(v: unknown): v is SecretRef {
  return typeof v === "object" && v !== null && (v as Record<symbol, unknown>)[SECRET_BRAND] === true;
}

/**
 * `secrets["NAME"]` returns a SecretRef — a reference resolved at run time
 * (locally from .env + the process env; in the cloud from the secret store).
 */
export const secrets: Readonly<Record<string, SecretRef>> = new Proxy(
  {},
  {
    get(target, prop: string | symbol): SecretRef | undefined {
      // Only string subscripting names a secret; symbol access defers to the
      // target (so the object is not an accidental thenable, etc.).
      if (typeof prop === "symbol") {
        return (target as Record<symbol, never>)[prop];
      }
      const name = prop;
      if (!SECRET_NAME.test(name)) {
        throw new Error(
          `secret name "${name}" is invalid: names must match [A-Za-z_][A-Za-z0-9_]* ` +
            `(letters, digits, underscores; no leading digit).`,
        );
      }
      return { [SECRET_BRAND]: true, name };
    },
  },
) as Readonly<Record<string, SecretRef>>;
